### DISCUSSION TITLE
How to import dependency for the functions?

### DISCUSSION THREAD

**Discussion by k82cn** [PyO3/pyo3#3862] [Questions] [answered]
I'd like to run following Python code in Rust, but I'm not sure how to import its dependencies, e.g. random, math.

```python
def pi(n):
    sum = 0.0

    for i in range(n):
        x, y = random.uniform(0, 1.0), random.uniform(0, 1.0)
        if math.hypot(x, y) <= 1.0:
            sum += 1

    return sum
```

During execution, I got the following errors; so how to import dependencies for functions.

```
Error: PyErr { type: <class 'NameError'>, value: NameError("name 'random' is not defined"), traceback: Some(<traceback object at 0x7f3bc81a9e40>) }
```

**Comment by k82cn** [comment]
xref: https://github.com/k82cn/rust-learning/tree/main/pyo3

**Comment by adamreichold** [comment]
Did you try just including the relevant `import` statements as part of the code snippets?

**Reply by k82cn** [reply]
It's not my use case; I'll use dill to dump the function into object. And those dependencies are global instead of handling it in functions.

**Reply by adamreichold** [reply]
Then you will have to manually import the modules using [`Python::import`](https://docs.rs/pyo3/latest/pyo3/marker/struct.Python.html#method.import) and modify the globals, most likely by also importing `__main__` and accessing its dictionary via [`PyModule::dict`](https://docs.rs/pyo3/latest/pyo3/types/struct.PyModule.html#method.dict).

**Reply by k82cn** [reply]
It does not work after adding those two lines.

```
        println!("Execute task");
        Python::import(py, "random");
        Python::import(py, "math");
        let _ = task_fn.call(py, args, Some(kwargs))?;
```

**Reply by adamreichold** [reply]
As written above, you have to actually put the modules returned by `Python::import` into the globals dictionary, i.e. the module dictionary of `__main__`.

**Reply by k82cn** [reply]
Can you share some example?

**Reply by adamreichold** [reply, marked as answer]
I did not compile or test this, but it should go something like

```rust
let main = py.import("__main__")?;
let globals = main.dict();

let random = py.import("random")?;
globals.set_item("random", random)?;

let math = py.import("math")?;
globals.set_item("math", math)?;
```

Note that this is in no way scoped to the current invocation of `task_fn`, it will affect the global state of the embedded Python interpreter. (But your task function will not be isolated anyway if they share that single embedded interpreter.)

**Reply by k82cn** [reply]
It works, thanks very much.

**Reply by k82cn** [reply]
In addition, is there anyway to re-create a new interpreter without re-create the process?

**Reply by adamreichold** [reply]
> In addition, is there anyway to re-create a new interpreter without re-create the process?

IIRC, we cannot support this safely for now, see the docs of [`with_embedded_python_interpreter`](In addition, is there anyway to re-create a new interpreter without re-create the process?) as to why.

**Reply by k82cn** [reply]
Got that :) I will start a subprocess to handle that.

BTW, do you know how to translate `Py<Any>` to Bytes ( https://crates.io/crates/bytes )?

I'm going to dump this return value of `task_fn` by `dill.dumps` and transfer those data back to client through gRPC's bytes.

**Reply by davidhewitt** [reply]
There was discussion of `bytes::Bytes` in https://github.com/PyO3/pyo3/issues/1992 - no final decision was taken, but the advantage was unclear at the time.

**Reply by k82cn** [reply]
IMO, it's a bit different; I'll get data from `dill.dumps`, what I need is to transfer it into `Bytes`. Something like that.

```
        let dill = py.import("dill")?;
        let dump_fn = dill.getattr("dumps")?;
        let res = PyTuple::new(py, &[res]);

        // How to transfer to Bytes for those return value?
        let _: Py<PyAny> = dump_fn.call(res, None)?.into();
```
